Fixes Curve25519/Ed25519 for use in OpenPGP.js

mahrud opened the pull request:
This pull request fixes point encoding problems for X25519 curves, fixes key generation for Curve25519, and adds a key generation function for Ed25519. These are necessary fixes needed for adding ECC support to OpenPGP.js in this pull request: https://github.com/openpgpjs/openpgpjs/pull/618 .

mahrud commented:
@indutny have you had a chance to review this?
I didn't add any tests here since it still doesn't produce test vectors from #142, but there are new tests in the `ecc_nist` branch of OpenPGP.js that incorporate changes this PR.

indutny commented on lib/elliptic/ec/key.js:
```diff
@@ -51,13 +52,7 @@ KeyPair.prototype.validate = function validate() {
-    compact = null;
-  }
-
+KeyPair.prototype.getPublic = function getPublic(enc, compact) {
```
This is going to be a major semver change.

mahrud replied:
Do you know when you'd like to release v7.0.0? I can think of some other possibly breaking cleanups that might be worthwhile to do.

For instance, I think it would make sense to remove dependency on `hmac-drbg` and `brorand` and instead pass a callback function that returns randomness. This one is definitely breaking, but there are some other things that I can't quite remember off the top of my head.

chjj replied:
Don't know about this one. This breaks node.js convention: `enc` is typically the last argument. It's also going to be a huge pain replacing all the code which relies on the old parameter order.

Agree with removing the brorand dependency though.

mahrud replied:
I think consistency is more important here. Also, I've done the work of replacing the code which relies on the old order in this project (which was minimal), and other projects can easily fix it whenever they update the vendored version of elliptic.

indutny commented on lib/elliptic/eddsa/index.js:
```diff
@@ -74,6 +79,23 @@ EDDSA.prototype.keyFromSecret = function keyFromSecret(secret) {
+    nonce: this.curve.n.toArray()
+  });
+
+  return this.keyFromSecret(drbg.generate(32));
```
Let's not hard-code `32` here.

indutny commented on lib/elliptic/eddsa/key.js:
```diff
@@ -17,11 +17,18 @@ var cachedProperty = utils.cachedProperty;
+  else {
     this._pubBytes = parseBytes(params.pub);
+    if (this._pubBytes && this._pubBytes.length === 33 &&
+        this._pubBytes[0] === 0x40)
```
Please add `{` `}`.

mahrud replied:
Only here?

indutny replied:
Yeah, current code style allows omitting braces if the `if` statement and its body fits into two lines total (the body must always be on the next line).

dkg commented:
what's the status on this pull request?  it'd be nice to get it included so that openpgp.js doesn't use a fork :)

mahrud commented:
@dkg sorry about the delay. I think it's ready.

mahrud commented:
Is there anything else I should do here?

chesnokovilya commented:
@indutny Hey! So what prevents this PR from merging? Is it order of parameters in the line? `KeyPair.prototype.getPublic = function getPublic(enc, compact)`
I think in general brainpool and curve25519 support could be good for this library!

mahrud commented:
@chjj, I haven't used or followed the changes to this repository since last April, so I'm not confident I can add the conversion method you requested. I do remember a lot of pain dealing with conflicting endian-ness in different RFCs, and I kept that in mind when making this PR and added notes or references to relevant RFCs when necessary.

With regards to the order of `enc` and `compact`, my opinion is that checking the type of the first input and switching the order of inputs if it is a string, as it is currently, should be avoided.
https://github.com/indutny/elliptic/blob/ee7970b92f388e981d694be0436c4c8036b5d36c/lib/elliptic/ec/key.js#L53-L58
That said, if it is the only thing keeping this PR from being merged, then please feel free to amend the PR as you see fit.

I rebased the PR and brought it up to date with the version that openpgp.js uses. All tests are green.

mahrud commented:
Should I close this PR? Not sure why it's still open, despite being approved. If the changes have already been applied elsewhere, please feel free to close it.

mahrud commented:
I don't think so. It just was never merged...

Thijs-van-Drongelen commented:
Brainpool support would be nice
